src/target/riscv/riscv-013.c

```diff
@@ -582,65 +584,73 @@ static int dmi_op_timeout(struct target *target, uint32_t *data_in,
 	}
 
 	time_t start = time(NULL);
+	/* This first loop performs the request.  Note that if for some reason this
+	 * stays busy, it is actually due to the previous access. */
 	while (1) {
-		/* This first loop performs the request.  Note that if for some reason this
-		 * stays busy, it is actually due to the previous access. */
+		status = dmi_scan(target, NULL, NULL, dmi_op, address, data_out,
+				exec);
+		if (status == DMI_STATUS_BUSY) {
+			increase_dmi_busy_delay(target);
+			if (dmi_busy_encountered)
+				*dmi_busy_encountered = true;
+		} else if (status == DMI_STATUS_SUCCESS) {
+			break;
+		} else {
+			LOG_ERROR("failed %s at 0x%x, status=%d", op_name, address, status);
+			return ERROR_FAIL;
+		}
+		if (time(NULL) - start > timeout_sec)
+			return ERROR_TIMEOUT_REACHED;
+	}
+
+	if (status != DMI_STATUS_SUCCESS) {
+		LOG_ERROR("Failed %s at 0x%x; status=%d", op_name, address, status);
+		return ERROR_FAIL;
+	}
+
+	if (ensure_success) {
+		/* This second loop ensures the request succeeded, and gets back data.
+		 * Note that NOP can result in a 'busy' result as well, but that would be
+		 * noticed on the next DMI access we do. */
 		while (1) {
-			status = dmi_scan(target, NULL, NULL, dmi_op, address, data_out,
-					exec);
+			status = dmi_scan(target, &address_in, data_in, DMI_OP_NOP, address, 0,
+					false);
 			if (status == DMI_STATUS_BUSY) {
 				increase_dmi_busy_delay(target);
-				if (!retry)
-					return ERROR_TARGET_BUSY;
+				if (dmi_busy_encountered)
+					*dmi_busy_encountered = true;
 			} else if (status == DMI_STATUS_SUCCESS) {
 				break;
 			} else {
-				LOG_ERROR("failed %s at 0x%x, status=%d", op_name, address, status);
+				LOG_ERROR("failed %s (NOP) at 0x%x, status=%d", op_name, address,
+						status);
 				return ERROR_FAIL;
 			}
 			if (time(NULL) - start > timeout_sec)
 				return ERROR_TIMEOUT_REACHED;
-		}
 
-		if (status != DMI_STATUS_SUCCESS) {
-			LOG_ERROR("Failed %s at 0x%x; status=%d", op_name, address, status);
-			return ERROR_FAIL;
-		}
-
-		if (ensure_success) {
-			/* This second loop ensures the request succeeded, and gets back data.
-			 * Note that NOP can result in a 'busy' result as well, but that would be
-			 * noticed on the next DMI access we do. */
-			while (1) {
-				status = dmi_scan(target, &address_in, data_in, DMI_OP_NOP, address, 0,
-						false);
-				if (status == DMI_STATUS_BUSY) {
-					increase_dmi_busy_delay(target);
-					if (!retry)
-						return ERROR_TARGET_BUSY;
-					break;
-				} else if (status == DMI_STATUS_SUCCESS) {
-					return ERROR_OK;
+			if (status == DMI_STATUS_FAILED) {
+				if (data_in) {
```

Comment:
Is the body of this "if" statement reachable?  I see the assignment to status in line 616, and if the value DMI_STATUS_FAILED gets assigned to status at line 616, then the "else" clause at line 624 will be executed, including the return statement on line 627, in which case the status variable seems guaranteed to be unequal to DMI_STATUS_FAILED at line 632.

Reply:
You're right. Looks like some code got duplicated there. Thanks!